### ISSUE TITLE
[BUG] Huge difference between Rust optimizer and Python optimizer

### ISSUE THREAD

**Issue by L-M-Sherlock** [open-spaced-repetition/fsrs-rs#78]
The weights generated by Anki are very different from those generated by the python optimizer.

- Weights generated by Anki: 1.1408, 2.1175, 9.0984, 90.9615, 4.7957, 1.1697, 1.2771, 0.0010, 1.6411, 0.1000, 1.0769, 2.1627, 0.0928, 0.4849, 1.0269, 0.0000, 3.4083
- Weights generated by python optimizer: 1.1366, 1.4954, 20.8954, 41.342, 4.5015, 1.6955, 2.114, 0.0, 1.7707, 0.1551, 1.1898, 1.5944, 0.1889, 0.6854, 0.0115, 0.0, 6.6447
- Deck to reproduce: [Default.zip](https://github.com/ankitects/anki/files/12708324/Default.zip) (change file extension to .apkg)
- Timezone (for Python optimizer): Asia/Calcutta
- Other settings are Default

This issue is not caused by the bug where Anki calculates weights for all the reviews (not respecting the preset).

_Originally posted by @user1823 in https://github.com/ankitects/anki/issues/2443#issuecomment-1732545574_

**Comment by L-M-Sherlock** [comment]
I forget to remove outliers in the Rust optimizer:

```python
        def remove_outliers(group: pd.DataFrame) -> pd.DataFrame:
            grouped_group = (
                group.groupby(by=["r_history", "delta_t"], group_keys=False)
                .agg({"y": ["mean", "count"]})
                .reset_index()
            )
            sort_index = grouped_group.sort_values(
                by=[("y", "count"), "delta_t"], ascending=[True, False]
            ).index

            total = sum(grouped_group[("y", "count")])
            has_been_removed = 0
            for i in sort_index:
                count = grouped_group.loc[i, ("y", "count")]
                if has_been_removed + count >= total * 0.05:
                    break
                has_been_removed += count
            group = group[
                group["delta_t"].isin(
                    grouped_group[grouped_group[("y", "count")] >= count]["delta_t"]
                )
            ]
            return group

        df[df["i"] == 2] = (
            df[df["i"] == 2]
            .groupby(by=["r_history", "t_history"], as_index=False, group_keys=False)
            .apply(remove_outliers)
        )
        df.dropna(inplace=True)

        def remove_non_continuous_rows(group):
            discontinuity = group["i"].diff().fillna(1).ne(1)
            if not discontinuity.any():
                return group
            else:
                first_non_continuous_index = discontinuity.idxmax()
                return group.loc[: first_non_continuous_index - 1]

        df = df.groupby("card_id", as_index=False, group_keys=False).progress_apply(
            remove_non_continuous_rows
        )
```

**Comment by L-M-Sherlock** [comment]
fsrs-rs:
<img width="260" alt="image" src="https://github.com/open-spaced-repetition/fsrs-rs/assets/32575846/3097efe5-5904-4ee3-b49e-abd38f9b9f7b">
fsrs-py:
<img width="260" alt="image" src="https://github.com/open-spaced-repetition/fsrs-rs/assets/32575846/72608a3d-e45a-458d-ae22-df94d9144a26">

**Comment by L-M-Sherlock** [comment]
To increase `num_epochs` could reduce the errors. But it also will slow down the the optimization.

**Comment by L-M-Sherlock** [comment]
The left is generated by Anki. The right is generated by python optimizer.

![image](https://github.com/open-spaced-repetition/fsrs-rs/assets/32575846/fb6803ae-ff84-4261-a381-d2fdde26cc15)

**Comment by Expertium** [comment]
> To increase `num_epochs` could reduce the errors. But it also will slow down the the optimization.

By the way, how many epochs does the optimizer in the beta version use? Also, does it use splits, with averaging of the parameters afterwards?

**Comment by L-M-Sherlock** [comment]
> By the way, how many epochs does the optimizer in the beta version use? Also, does it use splits, with averaging of the parameters afterwards?

It uses 16 epochs and doesn't have splits because the framework doesn't support splits.

**Comment by Expertium** [comment]
So the code related to removing outliers will be added in the next release?

**Comment by L-M-Sherlock** [comment]
@asukaminato0721 and I will deal with it.

**Comment by L-M-Sherlock** [comment]
I implement the outlier filter, but the first four weights are still very different from the Python optimizer. So it's not caused by outlier filter. By the way, I think the weights generated by the Python optimizer doesn't fit the forgetting curve well:

<img width="595" alt="image" src="https://github.com/open-spaced-repetition/fsrs-rs/assets/32575846/86087d96-ac9c-45ec-90a2-f25a5fa04f0c">

A smaller value of stability would be better:

<img width="592" alt="image" src="https://github.com/open-spaced-repetition/fsrs-rs/assets/32575846/1229f706-f5d8-42bc-bcee-723b516f4856">

Maybe RMSE is not a good loss function here. I plan to use log loss.

**Comment by Expertium** [comment]
> Maybe RMSE is not a good loss function here. I plan to use log loss.

I would recommend running the benchmark with both RMSE and logloss to determine whether there is a difference in the final RMSE.

**Comment by user1823** [comment]
> I implement the outlier filter, but the first four weights are still very different from the Python optimizer. So it's not caused by outlier filter. By the way, I think the weights generated by the Python optimizer doesn't fit the forgetting curve well:

The supposed poor fitting of weights produced by the Python optimizer is definitely worth investigating. But, for now, the main focus should be on finding out why the first four weights generated by the Rust and the Python optimizer very different.

**Comment by Expertium** [comment]
> Maybe RMSE is not a good loss function here. I plan to use log loss.

I know this issue is closed, but I'm curious, did you end up testing RMSE vs logloss in pretrain? If so, which one is better?

**Comment by L-M-Sherlock** [comment]
I tested it. The log loss is more robust than RMSE in pretrain.
